translator/GCP_cloud_function/cloud_function/main.py

```diff
@@ -42,28 +47,73 @@ def translate_newest_icone_to_wzdx(event, context):
   return
 
 def get_ftp_file(url) :
-  with closing(request.urlopen(url)) as r:
-    return r.read().decode('utf-8-sig')
-
+  return urllib.request.urlopen(url).read().decode('utf-8-sig')
 
 def parse_xml(xml_string):
   return xmltodict.parse(xml_string)
 
 def get_wzdx_schema(schema_file_name):
-  return json.loads(open(schema_file_name,'r').read())
+  try:
+    schema_string=open(schema_file_name,'r').read()
+  except FileNotFoundError as e:
+    raise RuntimeError('invalid schema: file does not exist')
+  if not schema_string:
+    raise RuntimeError('invalid schema: empty schema')
+  try:
+    schema_obj= json.loads(schema_string)
+    jsonschema.validate('', schema_obj)  
+  except json.decoder.JSONDecodeError as e :
+    raise RuntimeError('invalid schema: not valid json') from e
+  except jsonschema.SchemaError as e :
+    raise RuntimeError('invalid schema: schema failed validation') from e
+  except jsonschema.ValidationError:
+    return schema_obj
+  return schema_obj
 
-#this function retrieves username and password info from secrets .This will be implemented after correct secrets are implemented.
 def get_ftp_credentials():
+
   secret_client = secretmanager.SecretManagerServiceClient()
-  username_secret_name = os.environ['icone_ftp_username_secret_name']
-  password_secret_name = os.environ['icone_ftp_password_secret_name']
-  project_id = os.environ['project_id']
+  username_secret_name = os.environ.get('icone_ftp_username_secret_name')
+  password_secret_name = os.environ.get('icone_ftp_password_secret_name')
+  project_id = os.environ.get('project_id')
+           
+  if not username_secret_name or not password_secret_name or not project_id:
+    return None
   request = {"name": f"projects/{project_id}/secrets/{username_secret_name}/versions/latest"}
-  response = client.access_secret_version(request)
+  try:
+    response = secret_client.access_secret_version(request)
+  except:
+    logging.error(RuntimeError('unable to get username from secret. Either the secret does not exist or there are no permissions!'))
+    return None, None
   username = response.payload.data.decode("UTF-8")
 
   request = {"name": f"projects/{project_id}/secrets/{password_secret_name}/versions/latest"}
-  response = client.access_secret_version(request)
+  try:
+    response = secret_client.access_secret_version(request)
+  except:
+    logging.error(RuntimeError('unable to get password from secret. Either the secret does not exist or there are no permissions!'))
+    return None, None
   password = response.payload.data.decode("UTF-8")
 
-  return username,password
\ No newline at end of file
+  return username,password
+
+
+def unsupported_messages_callback(message):
+  project_id = os.environ.get('project_id')
+  unsupported_messages_topic_id = os.environ.get('unsupported_messages_topic_id')
+  if not project_id or not unsupported_messages_callback:
```

Comment:
Since we're already in the `unsupported_messages_callback` method, we don't need to check for it's existence here.

Reply:
should be fixed

Reply:
This method is still the same, and still checking for the existence of the method from which you are calling.

Reply:
It looks like he accidentally had unsupported_messages_callback instead unsupported_messages_topic_id in the if statement. I had him change that, to check if the topic_id exists instead of accidentally checking if the function exists